UM/View/RenderBatch.py

```diff
@@ -0,0 +1,297 @@
+    #
+    #   \param transform The transformation matrix to use for rendering the item.
+    #   \param mesh The mesh to render with the transform matrix.
+    #   \param uniforms A dict of additional uniform bindings to set when rendering the item.
+    #                   Note these are set specifically for this item.
+    def addItem(self, transform, mesh, uniforms = None):
+        self._items.append((transform, mesh, uniforms))
+
+    ##  Render the batch.
+    #
+    #   \param camera The camera to render from.
+    def render(self, camera):
+        self._shader.bind()
+
+        if self._backface_cull:
+            self._gl.glEnable(self._gl.GL_CULL_FACE)
+        else:
+            self._gl.glDisable(self._gl.GL_CULL_FACE)
+
+        if self._render_type == self.RenderType.Solid:
+            self._gl.glDisable(self._gl.GL_BLEND)
+            self._gl.glEnable(self._gl.GL_DEPTH_TEST)
+            self._gl.glDepthMask(self._gl.GL_TRUE)
+        if self._render_type == self.RenderType.Transparent:
+            self._gl.glDepthMask(self._gl.GL_FALSE)
+            self._gl.glEnable(self._gl.GL_BLEND)
+            self._gl.glBlendFunc(self._gl.GL_SRC_ALPHA, self._gl.GL_ONE_MINUS_SRC_ALPHA)
+        elif self._render_type == self.RenderType.Overlay:
+            self._gl.glDisable(self._gl.GL_DEPTH_TEST)
+
+        self._view_matrix = camera.getWorldTransformation().getInverse()
+        self._projection_matrix = camera.getProjectionMatrix()
+        self._view_projection_matrix = camera.getProjectionMatrix().multiply(self._view_matrix)
+
+        self._shader.updateBindings(
+            view_matrix = self._view_matrix,
+            projection_matrix = self._projection_matrix,
+            view_projection_matrix = self._view_projection_matrix,
+            view_position = camera.getWorldPosition(),
+            light_0_position = camera.getWorldPosition() + Vector(0, 50, 0)
+        )
+
+        for item in self._items:
+            self._renderItem(item)
+
+        self._shader.release()
+
+    def _renderItem(self, item):
+        transform = item[0]
+        mesh = item[1]
+        if not mesh:
+            return #Something went wrong, node has no mesh.
+
+        normal_matrix = None
+        if mesh.hasNormals():
+            normal_matrix = copy.deepcopy(transform)
+            normal_matrix.setRow(3, [0, 0, 0, 1])
+            normal_matrix.setColumn(3, [0, 0, 0, 1])
+            normal_matrix = normal_matrix.getInverse().getTransposed()
+
+        model_view_matrix = copy.deepcopy(transform).preMultiply(self._view_matrix)
+        model_view_projection_matrix = copy.deepcopy(transform).preMultiply(self._view_projection_matrix)
+
+        self._shader.updateBindings(
+            model_matrix = item[0],
+            normal_matrix = normal_matrix,
+            model_view_matrix = model_view_matrix,
+            model_view_projection_matrix = model_view_projection_matrix
+        )
+
+        if item[2] is not None:
```

Comment:
So item is a list? I generally don't like this kind of constructs; it gives really weird errors if the order of the list ever changes. Better to use dicts.